pkg/resourcevalidator/webhook.go

```diff
@@ -0,0 +1,241 @@
+	case admissionv1.Delete:
+		obj = req.OldObject
+	case admissionv1.Update:
+		return spv.HandleUpdate(req)
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("unsupported operation %s", req.Operation))
+	}
+
+	// Decode the shadow pod
+	shadowpod, err := spv.DecodeShadowPod(obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Check existence and get shadow pod origin Cluster ID label
+	spClusterID, found := shadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		return admission.Denied("missing origin Cluster ID label")
+	}
+
+	// Get ShadowPod Namespace
+	namespace := &corev1.Namespace{}
+	if err := spv.Client.Get(ctx, client.ObjectKey{Name: shadowpod.GetNamespace()}, namespace); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Get Cluster ID origin label of the namespace
+	nsClusterID, found := namespace.Labels[consts.RemoteClusterID]
+	if !found || nsClusterID != spClusterID {
+		webhooklog.Info(fmt.Sprintf("\t\tNamespace Cluster ID [ %s ] does not match the ShadowPod Cluster ID [ %s ]", nsClusterID, spClusterID))
+		return admission.Denied("Namespace Cluster ID label does not match the ShadowPod Cluster ID label")
+	}
+
+	if !spv.enableResourceValidation {
+		return admission.Allowed("")
+	}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		return spv.HandleCreate(ctx, req, shadowpod, nsClusterID, pointer.BoolDeref(req.DryRun, false))
+	case admissionv1.Delete:
+		return spv.HandleDelete(ctx, req, shadowpod, nsClusterID, pointer.BoolDeref(req.DryRun, false))
+	default:
+		return admission.Denied("Unsupported operation")
+	}
+}
+
+// HandleCreate is the function in charge of handling Creation requests.
+//nolint:gocritic // the signature of this method is imposed by controller runtime.
+func (spv *ShadowPodValidator) HandleCreate(ctx context.Context, req admission.Request, shadowpod *vkv1alpha1.ShadowPod,
+	clusterID string, dryRun bool) admission.Response {
+	spQuota := getQuotaFromShadowPod(shadowpod)
+
+	shadowpodlog.Info(fmt.Sprintf("\tShadowPod %s decoded: UID: %s - clusterID %s", shadowpod.Name, shadowpod.GetUID(), clusterID))
+
+	// Check existence and get resource offer by Cluster ID label
+
+	resourceoffer, err := liqogetters.GetResourceOfferByLabel(ctx, spv.Client, corev1.NamespaceAll,
+		liqolabels.LocalLabelSelector(clusterID))
+	//resourceoffer, err := spv.getResourceOfferByLabel(ctx, clusterID)
+	if err != nil {
+		error := fmt.Errorf("error getting resource offer by label: %w", err)
+		resourceofferlog.Info(error.Error())
+		return admission.Errored(http.StatusInternalServerError, error)
+	}
+
+	// Get ResourceOffer Quota
+	roQuota := getQuotaFromResourceOffer(resourceoffer)
+
+	resourceofferlog.Info(fmt.Sprintf("ResourceOffer found for clusterID %s with %s ", clusterID, quotaFormatter(roQuota, "Quota")))
+
+	peeringInfo := getOrCreatePeeringInfo(spv.PeeringCache, clusterID, resourceoffer.OwnerReferences[0].Name, roQuota)
```

Comment:
It is risky to extract the cluster name this way, since it might panic if the ownerreference is not present. Since it appears to me it is only used for logging, I would drop the cluster name part for the sake of simplicity (it would require some more refactoring to be integrated cleanly).

Reply:
internally it is used also to eventually create the PeeringInfo